eng/pipelines/templates/steps/build-artifacts.yml

```diff
@@ -83,13 +90,42 @@ steps:
 
   - ${{ parameters.BeforePublishSteps }}
 
-  - publish: $(Build.ArtifactStagingDirectory)
-    artifact: packages 
-    condition: succeededOrFailed()
+  - pwsh: |
+      # handle where the parameter is undefined
+      $shouldPublishDocs = $true
+      $docParam = "${{ parameters.BuildDocs }}"
+      if ($docParam) {
+        # if we set it to false handle it here
+        $shouldPublishDocs = 	[System.Convert]::ToBoolean($docParam)
+      }
+      
+      $packagePublishes = ((Get-ChildItem '$(Build.ArtifactStagingDirectory)') | Measure-Object).Count
+      
+      if ($packagePublishes -eq 0) {
+          Write-Host "We are expecting to publish package artifacts, but no artifacts were found."
+          Write-Host (Get-ChildItem '$(Build.ArtifactStagingDirectory)')
+          exit 1
+      }
+
+      if ($shouldPublishDocs) {
+        $docsPublishes = ((Get-ChildItem '$(Build.SourcesDirectory)/_docs') | Measure-Object).Count
+        if ($docsPublishes -eq 0) {
+          Write-Host "We are expecting to publish documentation artifacts, but no documentation artifacts were found."
+          Write-Host (Get-ChildItem '$(Build.SourcesDirectory)/_docs')
+          exit 1
+        }
+      }
+    displayName: 'Verify Presence of Artifacts'
 
   - task: PublishBuildArtifacts@1
-    condition: and(succeededOrFailed(), ${{parameters.BuildDocs}})
-    displayName: 'Publish Artifacts'
+    displayName: 'Publish Package Artifacts'
+    inputs:
+      artifactName: 'packages'
+      pathtoPublish: '$(Build.ArtifactStagingDirectory)'
+
+  - task: PublishBuildArtifacts@1
+    condition: and(succeeded(), ${{ parameters.BuildDocs }})
```

Comment:
Do you want to skip this on all failures are only if there are no doc files?

Reply:
We only want to skip if we deliberately shut off docs. Otherwise, having no docs is a failure.